samtranslator/model/api/api_generator.py

```diff
@@ -603,6 +612,135 @@ def _construct_api_domain(  # noqa: PLR0912, PLR0915
+        if basepaths is None:
+            basepath_mapping = self._create_basepath_mapping(domain_name_arn, rest_api)
+            basepath_resource_list.extend([basepath_mapping])
+        else:
+            sam_expect(basepaths, self.logical_id, "Domain.BasePath").to_be_a_list_of(ExpectedType.STRING)
+            for basepath in basepaths:
+                # Remove possible leading and trailing '/' because a base path may only
+                # contain letters, numbers, and one of "$-_.+!*'()"
+                path = "".join(e for e in basepath if e.isalnum())
+                logical_id = "{}{}{}".format(self.logical_id, path, "BasePathMapping")
+                basepath_mapping = ApiGatewayBasePathMappingV2(
+                    logical_id, attributes=self.passthrough_resource_attributes
+                )
+                basepath_mapping.DomainNameArn = ref(domain_name_arn)
+                basepath_mapping.RestApiId = ref(rest_api.logical_id)
+                basepath_mapping.Stage = ref(rest_api.logical_id + ".Stage")
+                basepath_mapping.BasePath = path if normalize_basepath else basepath
+                basepath_resource_list.extend([basepath_mapping])
+
+        # Create the Route53 RecordSetGroup resource
+        record_set_group = None
+        route53 = self.domain.get("Route53")
+        if route53 is not None:
+            sam_expect(route53, self.logical_id, "Domain.Route53").to_be_a_map()
+            if route53.get("HostedZoneId") is None and route53.get("HostedZoneName") is None:
+                raise InvalidResourceException(
+                    self.logical_id,
+                    "HostedZoneId or HostedZoneName is required to enable Route53 support on Custom Domains.",
+                )
+
+            logical_id_suffix = LogicalIdGenerator(
+                "", route53.get("HostedZoneId") or route53.get("HostedZoneName")
+            ).gen()
+            logical_id = "RecordSetGroup" + logical_id_suffix
+
+            record_set_group = route53_record_set_groups.get(logical_id)
+
+            if route53.get("SeparateRecordSetGroup"):
+                sam_expect(
+                    route53.get("SeparateRecordSetGroup"), self.logical_id, "Domain.Route53.SeparateRecordSetGroup"
+                ).to_be_a_bool()
+                return ApiDomainResponseV2(
+                    domain,
+                    basepath_resource_list,
+                    self._construct_single_record_set_group(self.domain, domain_name, route53),
+                )
+
+            if not record_set_group:
+                record_set_group = self._get_record_set_group(logical_id, route53)
+                route53_record_set_groups[logical_id] = record_set_group
+
+            record_set_group.RecordSets += self._construct_record_sets_for_domain(self.domain, domain_name, route53)
+
+        return ApiDomainResponseV2(domain, basepath_resource_list, record_set_group)
```

Comment:
It pains me to see that all this code it's the same as the original `_construct_api_domain` except by returning `ApiDomainResponseV2` instead of V1 (and most of the full function really)

I would really like to just take this to another function that we can call in both places. 
- either a function that can returns either a v1 or a v2 resource
- a function where you pass the "Class" of the object that needs to be created.
- a function that returns "the parameters", and then you create the object outside.

Now the problem here is that the domain and basepath mapping resource can also be either v1 or v2, so it gets harder to get the types right.

Maybe we can chat about it to see if there's a good solution.

Reply:
Discussed offline that it's better to leave this as is for now. Will investigate further refactoring opportunities, but that will be part of another PR.